### PR TITLE
Adding duplicate field validation support.

### PR DESCRIPTION
**Pull request by bneradt**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

- Changes the _rules member variable to be a multimap so that it can support duplicate field verification. 
- Updates the Rules classes to support duplicate fields.
- Updates the YAML parser to support field value sequences as a way to express duplicate fields.

This address issue: https://github.com/yahoo/proxy-verifier/issues/5

### CODE REVIEW

File: local/include/core/ProxyVerifier.h [hunk 1]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
    * TextView for the rule to compare inputs to
    */
-  static std::shared_ptr<RuleCheck> make_absence(swoc::TextView name,
-                                                 swoc::TextView value);
+  static std::shared_ptr<RuleCheck> make_absence(
+      swoc::TextView name,
+      swoc::TextView value);
+
+  /**
+   * @param values (unusesd) The list of values specified in the YAML node.
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: "unusesd"? Also, don't pull lists out of YAML nodes as arguments - pass the node.

File: local/include/core/ProxyVerifier.h [hunk 2]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
+   */
+  static std::shared_ptr<RuleCheck> make_absence(
+      swoc::TextView name,
+      std::list<swoc::TextView> values);
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: By Turing's left toenail - pass a `std::list` BY VALUE? A taste of the lash for you, laddy.

File: local/include/core/ProxyVerifier.h [hunk 3]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
+   * @param name The name of the target field (null if not found).
+   * @param value The value of the target field (null if not found).
+   * @return Whether the check was successful.
    */
-  virtual bool test(swoc::TextView key, swoc::TextView name,
+  virtual bool test(swoc::TextView transaction_key, swoc::TextView name,
                     swoc::TextView value) const = 0;
+
+  virtual bool test(swoc::TextView transaction_key, swoc::TextView name,
+                    std::list<swoc::TextView> values) const = 0;
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: Gah.

Comment (SolidWallOfCode) [reply local/include/core/ProxyVerifier.h]: Looking back on this, I think it's an error to pull the `TextView` out before passing it in. Why not just pass the YAML node and let the method detect scalar vs. list? That's what I tend to do.

File: local/include/core/ProxyVerifier.h [hunk 4]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
    */
   EqualityCheck(swoc::TextView name, swoc::TextView value);
 
+  /** Construct @a EqualityCheck with a given name and set of expected values.
+   *
+   * @param name The name of the target field
+   * @param value The associated values with the target field,
+   * that is used with strcasecmp comparisons
+   */
+  EqualityCheck(swoc::TextView name, std::list<swoc::TextView> values);
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: _stroke_

File: local/include/core/ProxyVerifier.h [hunk 5]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
+   *
+   * Reports errors in verbose mode.
+   *
+   * @param key The identifying transaction key.
+   * @param name The name of the target field (null if not found)
+   * @param values The values of the target field (null if not found)
+   * @return Whether the check was successful or not
+   */
+  bool test(swoc::TextView key, swoc::TextView name,
+            std::list<swoc::TextView> values) const override;
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: _stroke_

File: local/include/core/ProxyVerifier.h [hunk 6]
```diff
@@ -202,148 +212,252 @@ class RuleCheck {
...
    */
   bool test(swoc::TextView key, swoc::TextView name,
             swoc::TextView value) const override;
+
+  /**
+   * @param values (unused) The value of the target field (null
+   * if not found)
+   */
+  bool test(swoc::TextView key, swoc::TextView name,
+            std::list<swoc::TextView> values) const override;
```
Comment (SolidWallOfCode) [inline local/include/core/ProxyVerifier.h]: _stroke_

File: local/src/core/ProxyVerifier.cc [hunk 7]
```diff
@@ -1284,6 +1350,38 @@ bool EqualityCheck::test(swoc::TextView key, swoc::TextView name,
...
 
+bool EqualityCheck::test(swoc::TextView key, swoc::TextView name,
+                         std::list<swoc::TextView> values) const {
+  swoc::Errata errata;
+  if (name.empty()) {
+    errata.info(
+        R"(Equals Violation: Absent. Key: "{}", Name: "{}", Correct Value: "{}")",
+        key, _name, _value);
+  } else if (_values != values) {
+    swoc::LocalBufferWriter<MAX_LOCAL_MESSAGE_SIZE> message;
```
Comment (SolidWallOfCode) [inline local/src/core/ProxyVerifier.cc]: This is generally better done via
```
using MSG_BUFF = swoc::LocalBufferWriter<1024>;
// ...
MSG_BUFF message;
message.print(...)
```

File: local/include/core/ProxyVerifier.h (hunk 2 above)
Comment (bneradt) [reply local/include/core/ProxyVerifier.h]: Wounds from a friend can be trusted.

File: local/src/core/ProxyVerifier.cc (hunk 7 above)
Comment (bneradt) [reply local/src/core/ProxyVerifier.cc]: Yeah, that looks good.

File: local/src/core/ProxyVerifier.cc [hunk 8]
```diff
@@ -1420,23 +1551,52 @@ HttpFields::parse_fields_and_rules(YAML::Node const &fields_rules_node,
...
+      // There's a list of values associated with this field. This
+      // indicates duplicate fields for the same field name.
+      std::list<TextView> values;
+      for (auto const &value: ValueNode) {
+        TextView localized_value{HttpHeader::localize(value.Scalar())};
+        values.emplace_back(localized_value);
+        _fields.emplace(name, localized_value);
+      }
+      if (node_size == 2 && assume_equality_rule) {
+        _rules.emplace(name, RuleCheck::make_equality(name, values));
```
Comment (SolidWallOfCode) [inline local/src/core/ProxyVerifier.cc]: You could consider storing the list in the field object, and making the `RuleCheck` depend on a reference or pointer to that.

Comment (SolidWallOfCode) [reply local/src/core/ProxyVerifier.cc]: Or move the `list` in to the equality checker.

File: local/src/core/ProxyVerifier.cc [hunk 9]
```diff
@@ -1663,22 +1823,46 @@ std::string HttpHeader::make_key() const {
...
-    if (found_iter == _fields_rules->_fields.cend()) {
-      if (!rule.second->test(key, swoc::TextView(), swoc::TextView())) {
-        issue_exists = true;
+  auto const& rules = rules_._rules;
+  auto const& fields = _fields_rules->_fields;
+  for (auto const &[name, rule_check] : rules) {
+    auto name_range = fields.equal_range(name);
+    auto field_iter = name_range.first;
+    if (rule_check->expects_duplicate_fields()) {
+      std::list<TextView> values;
```
Comment (SolidWallOfCode) [inline local/src/core/ProxyVerifier.cc]: Why is this declared and passed in? Why isn't it a local variable in `test`?

File: local/include/core/ProxyVerifier.h (hunk 3 above)
Comment (bneradt) [reply local/include/core/ProxyVerifier.h]: I see what you're saying, but that won't work here actually. These parameters to test() come in off the wire, not from the YAML file.

File: local/src/core/ProxyVerifier.cc (hunk 9 above)
Comment (bneradt) [reply local/src/core/ProxyVerifier.cc]: It's initialized here as empty and passed in empty, intentionally, in the case where there are no fields. But in the case where there are fields it is populated.

As discussed offline, I'll instead explicitly pass in an empty list when the list is empty and push this declaration down to the place where there are fields and it is populated.

**Pull request merged**